src/mon/OSDMonitor.cc

```diff
@@ -2668,6 +2673,56 @@ bool OSDMonitor::prepare_pg_created(MonOpRequestRef op)
   return true;
 }
 
+bool OSDMonitor::preprocess_pg_ready_to_merge(MonOpRequestRef op)
+{
+  op->mark_osdmon_event(__func__);
+  auto m = static_cast<MOSDPGReadyToMerge*>(op->get_req());
+  dout(10) << __func__ << " " << *m << dendl;
+  auto session = m->get_session();
+  if (!session) {
+    dout(10) << __func__ << ": no monitor session!" << dendl;
+    return true;
+  }
+  if (!session->is_capable("osd", MON_CAP_X)) {
+    derr << __func__ << " received from entity "
+         << "with insufficient privileges " << session->caps << dendl;
+    return true;
+  }
+  const pg_pool_t *pi = osdmap.get_pg_pool(m->pgid.pool());
+  if (!pi) {
+    derr << __func__ << " pool for " << m->pgid << " dne" << dendl;
+    return true;
+  }
+  if (pi->get_pg_num() <= m->pgid.ps()) {
+    dout(20) << " pg_num " << pi->get_pg_num() << " already < " << m->pgid << dendl;
+    return true;
+  }
+  if (pi->get_pg_num_pending() > m->pgid.ps()) {
+    dout(20) << " pg_num_pending " << pi->get_pg_num_pending() << " > " << m->pgid << dendl;
+    return true;
+  }
+  return false;
+}
+
+bool OSDMonitor::prepare_pg_ready_to_merge(MonOpRequestRef op)
+{
+  op->mark_osdmon_event(__func__);
+  auto m = static_cast<MOSDPGReadyToMerge*>(op->get_req());
+  dout(10) << __func__ << " " << *m << dendl;
+  pg_pool_t p;
+  if (pending_inc.new_pools.count(m->pgid.pool()))
+    p = pending_inc.new_pools[m->pgid.pool()];
+  else
+    p = *osdmap.get_pg_pool(m->pgid.pool());
+  assert(p.get_pg_num() == m->pgid.ps() + 1);
+  assert(p.get_pg_num_pending() <= m->pgid.ps());
+  p.dec_pg_num();
```

Comment:
<pre>
    if (n < (int)p.get_pg_num() - 1) {
      ss << "specified pg_num_pending " << n << " < pg_num (" << p.get_pg_num()
	 << ") - 1; only single pg decrease is currently supported";
      return -EINVAL;
    }
</pre>

Reply:
i made all of the osd code generic so that it doesn't care about the decrement-only rule so that it's only the monitor that (1) only allows you to decrease by 1 and (2) only understands ReadyToMerge messages that are pg_num - 1.

Reply:
Hmmm. I'm pretty uncomfortable with that; it means we're going to have outstanding untested code in the OSD (which I imagine we'll rely on to work in the future once we feel comfortable eliminating this requirement from the monitor?).

Reply:
is this changed in later commits, or testing of the multi-pg change in one map case added?

Reply:
It's always decreasing pg_num by one at a time.  The OSD code has to handle merging more than that because you might have an imported ancient pg from before and osdmap gap that covers mulitple merges.  There's no test case for that yet.